### DISCUSSION TITLE
self.signer.get_signature_mechanism(None).hash_algo

### DISCUSSION THREAD

**Discussion by Vishwamithra37** [MatthiasValvekens/pyHanko#115] [Q&A] [answered]
I am getting this error and don't know how to solve it :(

**Comment by MatthiasValvekens** [comment, marked as answer]
Sooo, I'm seeing a complaint about an error, but no stack trace... ;)

In all seriousness, `get_signature_mechanism(...).hash_algo` is called with `None` if you don't explicitly tell pyHanko what digest algorithm to use. PyHanko will try to derive the hash algorithm from the declared signature mechanism, but if that's also not explicitly specified, it will try to make a reasonable guess based on the signer certificate's public key. It's not particularly clever about that, though, and that will fail if there's no "obvious" choice.

Can you post a stack trace, preferably with some sample code?

**Reply by Vishwamithra37** [reply]
My intention:
Was to use BytesIO objects in place of file objects. Files are slow and hard to manage. 
This error occured when I tried to push the BytesIO object into the signer.load method. (After doing what I did in the pull request).

Stacktrace:
I cannot show that since, it had been resolved. Like, I'm currently using a makeshift solution for the whole issue using tempfiles. So that I won't be fighting with pyHanko design at every corner using the BytesIO objects. 

I kind of want pyHanko to do it all and not be limited to pure signing like, having an easy layer to access the certificates. Which as you rightly pointed out is already done by asn1crypto. BUT
THERE was no sane way to get the certificate details out except the human_freindly thing. Which is a string. (I WANTED A DICT). 
SO yup thanks :)
